server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java

```diff
@@ -749,20 +749,16 @@ public void deleteStaleSegments(int lastNMetadataFilesToKeep) throws IOException
             lastNMetadataFilesToKeep,
             sortedMetadataFileList.size()
         );
-        List<String> metadataFilesToBeDeleted = metadataFilesEligibleToDelete.stream().filter(metadataFile -> {
-            try {
-                return !isLockAcquired(metadataFile);
-            } catch (IOException e) {
-                logger.error(
-                    "skipping metadata file ("
-                        + metadataFile
-                        + ") deletion for this run,"
-                        + " as checking lock for metadata is failing with error: "
-                        + e
-                );
-                return false;
-            }
-        }).collect(Collectors.toList());
+        Set<String> allLockFiles;
+        try {
+            allLockFiles = new HashSet<>(mdLockManager.listAllLocks(MetadataFilenameUtils.METADATA_PREFIX));
+        } catch (Exception e) {
+            logger.error("Exception while fetching segment metadata lock files, skipping deleteStaleSegments", e);
+            return;
+        }
+        List<String> metadataFilesToBeDeleted = metadataFilesEligibleToDelete.stream()
+            .filter(metadataFile -> allLockFiles.contains(metadataFile) == false)
+            .collect(Collectors.toList());
```

Comment:
Not directly related to this PR - In the following code, we delete objects one by one. If we see failure in deleting a file, we move on to the next one. However, we don't delete the metadata file and it would be retried next time. On retry, we would try to delete all the files again. Is it possible to keep in memory this information and skip the deletion attempt in cases if the file has been deleted already. We might be able to save some remote usage quota on failures. 

https://github.com/opensearch-project/OpenSearch/blob/b0eab7579aaf7715c6eea7a3d1e206b7d1cb6760/server/src/main/java/org/opensearch/index/store/RemoteSegmentStoreDirectory.java#L784-L817

We can create an issue if this seems of value.

Reply:
https://github.com/opensearch-project/OpenSearch/issues/12019